### PR TITLE
[script.transmission] 1.0.0

### PR DESCRIPTION
**Pull request by basilgello**
### Description
The control panel add-on for Transmission Bittorrent client for Kodi 19.0 "Matrix" and newer versions

This repository is a rework of original Transmission-XBMC addon by Correl J. Roush for Kodi 19.0 "Matrix".

Since version 19.0 "Matrix", the Python 2 support will be dropped and only Python 3 addons will be supported.

Also, the original addon featured the 'Search Torrent' functionality using pirate sites. This violates the rules of official Kodl respository, so this addon has only 'Add Torrent File' functionality.
### Checklist:
- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: script.transmission/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+Transmission-XBMC
```
Comment (enen92) [inline script.transmission/README.md:1]: please change to Transmission for Kodi (this naming is not allowed as it violates our trademark)

File: script.transmission/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+=================
+
+A client for the popular [Transmission](http://www.transmissionbt.com/) BitTorrent application for [XBMC](http://xbmc.org/)
```
Comment (enen92) [inline script.transmission/README.md:4]: kodi?

File: script.transmission/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+Features
+--------
+
+Currently, Transmission-XBMC supports viewing, adding, removing, starting and stopping torrents. More advanced features may be added in future releases.
```
Comment (enen92) [inline script.transmission/README.md:9]: same here

File: script.transmission/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+### Screenshot
+![Screenshot](http://github.com/downloads/correl/Transmission-XBMC/transmission-xbmc.png)
+
+Installation
+------------
+
+This script is intended for Kodi 19 (Matrix).
+
+This script can be installed via the addon manager within XBMC. Attempting to install it manually may not work, as it requires the simplejson library which is automatically installed by the addon manager.
```
Comment (enen92) [inline script.transmission/README.md:19]: kodi

File: script.transmission/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+If your transmission daemon is running on a machine other than localhost, using a port other than 9091, or requires authentication, you will have to change the plugin settings before running it.
+
+License
+-------
+
+Transmission-XBMC is licensed under the terms of the [MIT license](http://www.opensource.org/licenses/mit-license.html).
```
Comment (enen92) [inline script.transmission/README.md:26]: same

File: script.transmission/addon.xml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+<addon id="script.transmission"
+       name="Transmission"
+       version="0.8.0"
+       provider-name="Correl Roush / Vasyl Gello">
+  <requires>
+    <import addon="xbmc.python" version="3.0.0" />
+    <import addon="script.module.beautifulsoup4" version="4.8.2" />
+    <import addon="script.module.kodi-six" version="0.1.3"/>
+    <import addon="script.module.requests" version="2.22.0"/>
+    <import addon="script.module.simplejson" version="3.17.0"/>
```
Comment (enen92) [inline script.transmission/addon.xml:11]: not needed anymore

File: script.transmission/README.md (hunk 2 above)
Comment (basilgello) [reply script.transmission/README.md:4]: Done

File: script.transmission/README.md (hunk 4 above)
Comment (basilgello) [reply script.transmission/README.md:19]: done

File: script.transmission/addon.xml (hunk 6 above)
Comment (basilgello) [reply script.transmission/addon.xml:11]: Removing these imports complaints on missing kodi_six. Or do you mean kodi-six should not be used here?

File: script.transmission/README.md (hunk 5 above)
Comment (basilgello) [reply script.transmission/README.md:26]: done

File: script.transmission/README.md (hunk 3 above)
Comment (basilgello) [reply script.transmission/README.md:9]: done

File: script.transmission/addon.xml (hunk 6 above)
Comment (enen92) [reply script.transmission/addon.xml:11]: I mean any reference to simplejson, this was only for helix and below

File: script.transmission/default.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import os
+import sys
+from kodi_six import xbmc
+from kodi_six import xbmcaddon
+
+__settings__ = xbmcaddon.Addon(id='script.transmission')
+__language__ = __settings__.getLocalizedString
+
+BASE_RESOURCE_PATH = xbmc.translatePath( os.path.join( __settings__.getAddonInfo('path'), 'resources', 'lib' ) )
```
Comment (enen92) [inline script.transmission/default.py:12]: please use relative imports and not `sys.path.append`

File: script.transmission/default.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
Comment (enen92) [inline script.transmission/default.py:1]: this file should be as small as possible, only used to import your main codebase, propagate `sys.argv` and run. Reason is this is the only file that is not compiled to bytecode by kodi's python interpreter.

You should consider do something like:

```python
from sys import argv
from resources.lib import myscript

myscript.run(argv)
```

File: script.transmission/resources/lib/player.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import os
+import sys
+from kodi_six import xbmc, xbmcaddon, xbmcgui
+from six import iteritems
+import transmissionrpc
+
+__settings__ = xbmcaddon.Addon(id='script.transmission')
+
+BASE_RESOURCE_PATH = xbmc.translatePath( os.path.join( __settings__.getAddonInfo('path'), 'resources', 'lib' ) )
+sys.path.append (BASE_RESOURCE_PATH)
```
Comment (enen92) [inline script.transmission/resources/lib/player.py:13]: same

File: script.transmission/resources/lib/player.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+import common
+
+class SubstitutePlayer(xbmc.Player):
+    def __init__(self):
+        xbmc.Player.__init__(self)
+        self.prev_settings = {}
+        self.refreshSettings()
+
+    def onPlayBackStarted(self):
```
Comment (enen92) [inline script.transmission/resources/lib/player.py:23]: replaced by `onAvStarted` in leia

File: script.transmission/resources/lib/player.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        self.refreshSettings()
+        if self.active and xbmc.Player().isPlayingVideo():
```
Comment (enen92) [inline script.transmission/resources/lib/player.py:25]: It is not guaranteed that when this event is called (player initialization) player is already playingvideo, another reason to change it to onAVStarted

File: script.transmission/resources/lib/search.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+socket.setdefaulttimeout(15)
+
+class Search:
+    def __init__(self):
+        return NotImplemented
+    def search(terms):
+        return NotImplemented
+
+class TPB(Search):
```
Comment (enen92) [inline script.transmission/resources/lib/search.py:16]: Oh well, not allowed at all!

Comment (enen92) [reply script.transmission/resources/lib/search.py:16]: Drop any piracy related functionality from the addon if you want it on our repo.

File: script.transmission/resources/lib/search.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            td = td.findNext('td')
+            leechers = int(td.text)
+            torrents.append({
+                'url': url,
+                'name': name,
+                'seeds': seeds,
+                'leechers': leechers,
+            })
+        return torrents
+class Kickass(Search):
```
Comment (enen92) [inline script.transmission/resources/lib/search.py:46]: ...

File: script.transmission/resources/lib/search.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+                    'seeds': int(item.find('torrent:seeds').text),
+                    'leechers': int(item.find('torrent:peers').text),
+                })
+        except HTTPError as e:
+            if e.code == 404:
+                pass
+            else:
+                raise
+        return torrents
+class L337x(Search):
```
Comment (enen92) [inline script.transmission/resources/lib/search.py:69]: ...

File: script.transmission/resources/lib/search.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+            if not link:
+                continue
+            torrents.append({
+                'url': link['href'],
+                'name': details.text,
+                'seeds': seeds,
+                'leechers': leechers,
+            })
+        return torrents
+class Lime(Search):
```
Comment (enen92) [inline script.transmission/resources/lib/search.py:97]: ...

File: script.transmission/resources/skins/Default/skin.xml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (enen92) [inline script.transmission/resources/skins/Default/skin.xml:1]: remove this file from the submission, this is no longer used (and never was)

File: script.transmission/README.md (hunk 4 above)
Comment (enen92) [inline script.transmission/README.md:19]: remove simplejson

**Comment by enen92** [comment]
@basilgello for your information: https://github.com/xbmc/repo-scripts/pull/1455

Sorry, we have to do it. You're free to submit an update that drops the search functionality on piracy sites.

**Comment by basilgello** [comment]
> @basilgello for your information: #1455
> 
> Sorry, we have to do it. You're free to submit an update that drops the search functionality on piracy sites.

OK. I will drop the search functionality now.

File: script.transmission/README.md [hunk 17]
```diff
@@ -0,0 +1,26 @@
+Transmission-Kodi
```
Comment (enen92) [inline script.transmission/README.md:1]: transmission-kodi violates our trademark policy...

Please rename it to just Transmission, Transmission Client, or "Transmission for Kodi or whatever that doesn't cause confusion as the source.

File: script.transmission/README.md [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+This script is intended for Kodi 19 (Matrix).
+
+This script can be installed via the addon manager within Kodi. Attempting to install it manually may not work, as it requires the simplejson library which is automatically installed by the addon manager.
+
+If your transmission daemon is running on a machine other than localhost, using a port other than 9091, or requires authentication, you will have to change the plugin settings before running it.
+
+License
+-------
+
+Transmission-Kodi is licensed under the terms of the [MIT license](http://www.opensource.org/licenses/mit-license.html).
```
Comment (enen92) [inline script.transmission/README.md:26]: same here

File: script.transmission/README.md [hunk 19]
```diff
@@ -0,0 +1,26 @@
+Transmission-Kodi
+=================
+
+A client for the popular [Transmission](http://www.transmissionbt.com/) BitTorrent application for [Kodi](http://kodi.tv/)
+
+Features
+--------
+
+Currently, Transmission-Kodi supports viewing, adding, removing, starting and stopping torrents. More advanced features may be added in future releases.
```
Comment (enen92) [inline script.transmission/README.md:9]: same here

File: script.transmission/README.md (hunk 17 above)
Comment (basilgello) [reply script.transmission/README.md:1]: Transmission Control Panel sounds right?

File: script.transmission/addon.xml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    <summary lang="ru_RU">Панель управления для популярного торрент-клиента Transmission</summary>
+    <summary lang="uk_UA">Панель керування для популярного торент-кліента Transmission</summary>
+    <description lang="en_GB">Transmission-Kodi supports viewing, adding, removing, starting and stopping torrents.</description>
+    <description lang="hu_HU">Transmission-Kodi képes megnézni, hozzáadni, elkezdeni, és megállítani a torrenteket.</description>
+    <description lang="ca_ES">Transmission-Kodi permet veure, afegir, eliminar, iniciar i parar torrents.</description>
+    <description lang="ru_RU">Transmission-Kodi позволяет просматривать список торрентов, добавлять, удалять, запускать и останавливать торренты.</description>
+    <description lang="uk_UA">Transmission-Kodi дозволяє переглядати список торентів, додавати, видаляти, стартувати та зупиняти торенти.</description>
+    <platform>all</platform>
+    <license>MIT</license>
+    <forum></forum>
```
Comment (enen92) [inline script.transmission/addon.xml:30]: remove forum element if empty (this is parsed by our wiki and addon websites)

File: script.transmission/addon.xml [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+             library="default.py" />
+  <extension point="xbmc.service" 
+             library="resources/lib/player.py" />
+  <extension point="xbmc.addon.metadata">
+    <summary lang="en_GB">A client for the popular Transmission BitTorrent application</summary>
+    <summary lang="hu_HU">A népszerű Transmission Bittorrent alkalmazás vezérlése</summary>
+    <summary lang="ca_ES">Un client de Transmission, la popular aplicació de Bittorrent</summary>
+    <summary lang="ru_RU">Панель управления для популярного торрент-клиента Transmission</summary>
+    <summary lang="uk_UA">Панель керування для популярного торент-кліента Transmission</summary>
+    <description lang="en_GB">Transmission-Kodi supports viewing, adding, removing, starting and stopping torrents.</description>
```
Comment (enen92) [inline script.transmission/addon.xml:23]: remove Transmission-Kodi and use just transmission (or similar)

File: script.transmission/README.md (hunk 17 above)
Comment (enen92) [reply script.transmission/README.md:1]: Sure :)

**Comment by basilgello** [comment]
@enen92 I have refactored the code and tested it. Now adding torrent files works but piracy search is removed. Also I cleaned some dependencies and removed spaghetti code like sys.path manipulation.

Please have a look again and thanks for previous 20+ comments :)

**Pull request merged**